src/backend/columnar/columnar_customscan.c

```diff
@@ -501,16 +500,130 @@ RelationIdGetNumberOfAttributes(Oid relationId)
 }
 
 
-static Path *
-CreateColumnarScanPath(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
+/*
+ * AddColumnarScanPaths is the entry point for recursively generating
+ * parameterized paths. See AddColumnarScanPathsRec() for discussion.
+ */
+static void
+AddColumnarScanPaths(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 {
-	ColumnarScanPath *cspath = (ColumnarScanPath *) newNode(sizeof(ColumnarScanPath),
-															T_CustomPath);
+	Relids requiredRelids = rel->lateral_relids;
+	Relids remainingRelids = bms_difference(root->all_baserels, rel->relids);
 
 	/*
-	 * popuate custom path information
+	 * XXX: could avoid creating and discarding too many useless paths by
+	 * pre-filtering remainingRelids.
 	 */
-	CustomPath *cpath = &cspath->custom_path;
+
+	AddColumnarScanPathsRec(root, rel, rte, requiredRelids, remainingRelids,
+							-1);
+}
+
+
+/*
+ * AddColumnarScanPathsRec is a recursive function to search the
+ * parameterization space and add ColumnarScanPaths.
+ *
+ * Columnar tables resemble indexes because of the ability to push down
+ * quals. Ordinary quals, such as x = 7, can be pushed down easily. But join
+ * quals of the form "Var op Var" (where the relids of the two Vars are
+ * different) take more work and require the proper parameterization.
+ *
+ * Paths that require more outer rels can push down more join clauses that
+ * depend on those outer rels. But requiring outer rels gives the planner
+ * fewer options for the shape of the plan, so there is a trade-off. We always
+ * need to generate one minimally-parameterized path (parameterized only by
+ * lateral refs, if present) to make sure that at least one path can be
+ * chosen. Then, we generate as many useful parameterized paths as we
+ * reasonably can.
+ *
+ * The set of all possible parameterizations is the power set of
+ * remainingRelids. The power set has cardinality 2^N, where N is the
+ * cardinality of remainingRelids. XXX: do more to handle the worst cases.
+ */
+static bool
+AddColumnarScanPathsRec(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte,
+						Relids requiredRelids, Relids remainingRelids,
+						int prevNumClauses)
+{
+	CustomPath *customPath = CreateColumnarScanPath(root, rel, rte,
+													bms_copy(requiredRelids));
+
+	/* find the number of join clauses for this path */
+	int numClauses = 0;
+	if (customPath->path.param_info != NULL)
+	{
+		numClauses = list_length(customPath->path.param_info->ppi_clauses);
+	}
+
+	/*
+	 * Adding extra required relids should not decrease the number of join
+	 * clauses we can push down.
+	 */
+	Assert(numClauses >= prevNumClauses);
+
+	/*
+	 * If adding a new relid to the parameterization added new clauses, then
+	 * add the path and recurse. Otherwise discard it.
+	 */
+	if (numClauses == prevNumClauses)
+	{
+		pfree(customPath);
```

Comment:
we won't use the `customPath` here. However we have passed a copy of `requiredRelids` that is retained on the `customPath->param_info` under certain conditions.
 - is there a way to free a whole node structure?
 - would it be safe to free the whole customPath, or does it hold pointers to shared datastructures?
 - would it be feasible to free the `requiredRelids` copy?

Or do you estimate that the copies of `requiredRelids` are negligible in what they will leak here?
I am asking as the worst case of this function is described as `O(2^N)`, which escalates quickly.

I assume all memory is allocated in a Planner context which will be reset after planning ofc, so it might be ok to leak some of the copies here. In that case, what do we win by free-ing `customPath` here? Not to say to skip that, as freeing that is correct and easy. Just understanding the trade offs.

Reply:
I didn't see a clean way to do that. A few leaks here aren't necessarily too bad, they will get cleaned up after planning. But it's a good idea to be tighter about memory usage when searching a larger space (e.g. lots of parameterizations).

Ultimately, I removed the attempts to look at the Path after creation to see if it actually helps. Instead, I do more work upfront to avoid ever creating a useless path -- I only generate parameterizations for rels that are likely to be useful, that is, they take part in an expression that we are likely to be able to push down. That means we always add_path() for whatever we create, but we create fewer useless paths.